File: frontend/packages/core/src/model/stack.ts
import { produce } from "immer";
import { Entity } from "./entity";
import { createSignal, type Signal } from "../core/reactivity";
import { assert, assertInt } from "../core/assert";
import type { Point, StackId } from "./types";
import type { CardEntity } from "./card";

export class StackEntity extends Entity {
  pos: Signal<Point>;
  cards: Signal<CardEntity[]>;
  z: Signal<number>;

  constructor(id: StackId, initialPos: Point, initialCards: CardEntity[] = []) {
    super(id);
    this.pos = createSignal<Point>(initialPos);
    this.cards = createSignal<CardEntity[]>(initialCards);
    this.z = createSignal<number>(1);
  }

  // ---------- Access helpers ----------
  get size() {
    return this.cards[0]().length;
  }

  topIndex() {
    return this.size - 1;
  }

  topCard(): CardEntity | undefined {
    const cs = this.cards[0]();
    return cs[cs.length - 1];
  }

  bottomCard(): CardEntity | undefined {
    const cs = this.cards[0]();
    return cs[0];
  }

  // ---------- Mutations (Immer-backed) ----------
  setCards(next: CardEntity[]) {
    assert(Array.isArray(next), "StackEntity.setCards: next must be an array");
    this.cards[1](next);
  }

  /** Remove and return top card. Soft-fails on empty. */
  takeTop(): CardEntity | null {
    const before = this.cards[0]();
    if (before.length === 0) return null;

    let taken: CardEntity | null = null;
    this.cards[1](
      produce(before, (draft) => {
        taken = draft.pop() ?? null;
      })
    );

    // postcondition: length decreased by 1 and taken exists
    const after = this.cards[0]().length;
    assert(taken !== null, "StackEntity.takeTop: taken must not be null after pop");
    assert(after + 1 === before.length, "StackEntity.takeTop: size mismatch after pop");

    return taken;
  }

  /** Remove and return bottom card. Soft-fails on empty. */
  takeBottom(): CardEntity | null {
    const before = this.cards[0]();
    if (before.length === 0) return null;

    let taken: CardEntity | null = null;
    this.cards[1](
      produce(before, (draft) => {
        taken = draft.shift() ?? null;
      })
    );

    // postcondition: length decreased by 1 and taken exists
    const after = this.cards[0]().length;
    assert(taken !== null, "StackEntity.takeBottom: taken must not be null after shift");
    assert(after + 1 === before.length, "StackEntity.takeBottom: size mismatch after shift");

    return taken;
  }

  /**
   * Remove and return a range [start..end) (bottom-inclusive range in the array).
   * Soft-fails if out of bounds or invalid range.
   *
   * Hard asserts:
   * - indices must be integers
   */
  takeRange(start: number, endExclusive: number): CardEntity[] | null {
    assertInt(start, "StackEntity.takeRange(start)");
    assertInt(endExclusive, "StackEntity.takeRange(endExclusive)");

    const before = this.cards[0]();
    if (start < 0 || endExclusive > before.length || start >= endExclusive) return null;

    let out: CardEntity[] = [];
    this.cards[1](
      produce(before, (draft) => {
        out = draft.splice(start, endExclusive - start);
      })
    );

    // postcondition: removed exactly out.length cards
    const after = this.cards[0]().length;
    assert(after + out.length === before.length, "StackEntity.takeRange: size mismatch after splice");

    return out;
  }

  /**
   * Split stack in place at `index`:
   * [1,2,3,4,5], index=2 => this becomes [1,2], returns [3,4,5]
   *
   * Soft-fails for index out of range.
   * Hard asserts: index must be integer.
   */
  splitFrom(index: number): CardEntity[] | null {
    assertInt(index, "StackEntity.splitFrom(index)");

    const before = this.cards[0]();
    if (index <= 0 || index >= before.length) return null;

    let pulled: CardEntity[] = [];
    this.cards[1](
      produce(before, (draft) => {
        pulled = draft.splice(index);
      })
    );

    // postcondition: pulled length is before.length - index
    const after = this.cards[0]().length;
    assert(pulled.length === before.length - index, "StackEntity.splitFrom: pulled length mismatch");
    assert(after === index, "StackEntity.splitFrom: remaining size mismatch");

    return pulled;
  }

  /** Merge another stack's cards on top of this stack. No-op if other is empty. */
  mergeFrom(other: StackEntity) {
    const a = this.cards[0]();
    const b = other.cards[0]();
    if (b.length === 0) return;

    const beforeA = a.length;

    this.cards[1](
      produce(a, (draft) => {
        draft.push(...b);
      })
    );

    // postcondition: grew by b.length
    const afterA = this.cards[0]().length;
    assert(afterA === beforeA + b.length, "StackEntity.mergeFrom: size mismatch after push");
  }

  /** True if more than 1 card. */
  isStacked() {
    return this.size > 1;
  }

  /**
   * Used by "unstack" -> create N single-card stacks.
   * Returns arrays of 1 card each (does not create stacks; engine does that).
   */
  unstackIntoSingles(): CardEntity[][] {
    const cs = this.cards[0]();
    if (cs.length <= 1) return [cs.slice()];
    return cs.map((c) => [c]);
  }
}

File: internal/model/stack.go
package model

// StackID is a unique identifier for a stack.
type StackID string

// Point represents a 2D position.
type Point struct {
	X int `json:"x"`
	Y int `json:"y"`
}

// Stack is a collection of cards at a position on the board.
// Cards are ordered bottom-to-top (index 0 is bottom, last index is top).
type Stack struct {
	ID    StackID  `json:"id"`
	Pos   Point    `json:"pos"`
	Z     int      `json:"z"`
	Cards []CardID `json:"cards"`
}

// NewStack creates a new stack with the given ID and position.
func NewStack(id StackID, pos Point, cards []CardID) *Stack {
	if cards == nil {
		cards = []CardID{}
	}
	return &Stack{
		ID:    id,
		Pos:   pos,
		Z:     1,
		Cards: cards,
	}
}

// Size returns the number of cards in the stack.
func (s *Stack) Size() int {
	return len(s.Cards)
}

// TopIndex returns the index of the top card, or -1 if empty.
func (s *Stack) TopIndex() int {
	return len(s.Cards) - 1
}

// TopCard returns the ID of the top card, or empty string if empty.
func (s *Stack) TopCard() CardID {
	if len(s.Cards) == 0 {
		return ""
	}
	return s.Cards[len(s.Cards)-1]
}

// BottomCard returns the ID of the bottom card, or empty string if empty.
func (s *Stack) BottomCard() CardID {
	if len(s.Cards) == 0 {
		return ""
	}
	return s.Cards[0]
}

// IsStacked returns true if the stack has more than one card.
func (s *Stack) IsStacked() bool {
	return len(s.Cards) > 1
}

// TakeTop removes and returns the top card. Returns empty string if empty (soft failure).
func (s *Stack) TakeTop() CardID {
	if len(s.Cards) == 0 {
		return ""
	}
	top := s.Cards[len(s.Cards)-1]
	s.Cards = s.Cards[:len(s.Cards)-1]
	return top
}

// TakeBottom removes and returns the bottom card. Returns empty string if empty (soft failure).
func (s *Stack) TakeBottom() CardID {
	if len(s.Cards) == 0 {
		return ""
	}
	bottom := s.Cards[0]
	s.Cards = s.Cards[1:]
	return bottom
}

// TakeRange removes and returns cards in range [start, endExclusive).
// Returns nil if out of bounds or invalid range (soft failure).
func (s *Stack) TakeRange(start, endExclusive int) []CardID {
	if start < 0 || endExclusive > len(s.Cards) || start >= endExclusive {
		return nil
	}

	taken := make([]CardID, endExclusive-start)
	copy(taken, s.Cards[start:endExclusive])

	// Remove the range from cards
	s.Cards = append(s.Cards[:start], s.Cards[endExclusive:]...)

	return taken
}

// SplitFrom splits the stack at index, keeping [0..index) and returning [index..end].
// Returns nil if index out of range (soft failure).
// Example: [A,B,C,D,E], index=2 => stack becomes [A,B], returns [C,D,E]
func (s *Stack) SplitFrom(index int) []CardID {
	if index <= 0 || index >= len(s.Cards) {
		return nil
	}

	pulled := make([]CardID, len(s.Cards)-index)
	copy(pulled, s.Cards[index:])

	s.Cards = s.Cards[:index]

	return pulled
}

// MergeFrom appends all cards from another stack onto this stack.
// The other stack's cards go on top. No-op if other is empty.
func (s *Stack) MergeFrom(other *Stack) {
	if other == nil || len(other.Cards) == 0 {
		return
	}
	s.Cards = append(s.Cards, other.Cards...)
}

// UnstackIntoSingles returns each card as a separate slice (for creating individual stacks).
// Returns a slice of single-card slices.
func (s *Stack) UnstackIntoSingles() [][]CardID {
	if len(s.Cards) <= 1 {
		return [][]CardID{append([]CardID{}, s.Cards...)}
	}

	result := make([][]CardID, len(s.Cards))
	for i, card := range s.Cards {
		result[i] = []CardID{card}
	}
	return result
}

// SetCards replaces the card list.
func (s *Stack) SetCards(cards []CardID) {
	if cards == nil {
		cards = []CardID{}
	}
	s.Cards = cards
}
